File: lua/EventsApi.lua
local MakeRequest = require(script.Parent.HTTP).MakePlayFabApiCall
local PlayFab = require(script.Parent.PlayFab)
local Promise = require(script.Parent.Promise)
local TS = require(script.Parent.TS)

local PlayerData = PlayFab.PlayerData
local IsClientLoggedIn = PlayFab.IsClientLoggedIn

local WriteEvents = Promise.promisify(function(player, request)
	local data = PlayerData[player]
	if not IsClientLoggedIn(player) or not data then
		error("Must be logged in to call this method")
	end

	return TS.await(MakeRequest("/Event/WriteEvents", request, "X-EntityToken", data.EntityToken))
end)

local WriteTelemetryEvents = Promise.promisify(function(player, request)
	local data = PlayerData[player]
	if not IsClientLoggedIn(player) or not data then
		error("Must be logged in to call this method")
	end

	return TS.await(MakeRequest("/Event/WriteTelemetryEvents", request, "X-EntityToken", data.EntityToken))
end)

return {
	WriteEvents = WriteEvents,
	WriteTelemetryEvents = WriteTelemetryEvents,
}

File: src/EventsApi.ts
/// <reference path="EventsApi.d.ts" />

import MakeRequest from './HTTP'
import { PlayerData, IsClientLoggedIn } from './PlayFab'

/**
 * Write batches of entity based events to PlayStream. The namespace of the Event must be 'custom' or start with 'custom.'.
 * https://docs.microsoft.com/rest/api/playfab/events/playstream-events/writeevents
 */
export const WriteEvents = async (player: Player, request: WriteEventsRequest) => {
    const data = PlayerData.get(player)
    if ((!IsClientLoggedIn(player)) || (!data)) throw 'Must be logged in to call this method'
    const result = await MakeRequest<WriteEventsResponse>('/Event/WriteEvents', request, 'X-EntityToken', data.EntityToken)
        .catch((e: string) => { throw e })
    return result
}

/**
 * Write batches of entity based events to as Telemetry events (bypass PlayStream). The namespace must be 'custom' or start
 * with 'custom.'
 * https://docs.microsoft.com/rest/api/playfab/events/playstream-events/writetelemetryevents
 */
export const WriteTelemetryEvents = async (player: Player, request: WriteEventsRequest) => {
    const data = PlayerData.get(player)
    if ((!IsClientLoggedIn(player)) || (!data)) throw 'Must be logged in to call this method'
    const result = await MakeRequest<WriteEventsResponse>('/Event/WriteTelemetryEvents', request, 'X-EntityToken', data.EntityToken)
        .catch((e: string) => { throw e })
    return result
}